src/caffe/layers/cudnn_conv_layer.cpp

```diff
@@ -297,84 +124,103 @@ void CuDNNConvolutionLayer<Dtype>::Reshape(
+
+      // choose the fastest within limit
+      // if all algorithms exceed memory limit, we will use the 0 algorithm with no workspace
+      for (int a = 0; a < kRequestedForwardAlgoCount; ++a){
+        if ((fwd_perf[a].memory * this->group_ / 1024 / 1024) < Caffe::cudnn_mem_richness()
+            || Caffe::cudnn_mem_richness() == 0){
+          fwd_algo_[i] = fwd_perf[a].algo;
+          break;
+        }
+      }
 
-    // choose forward and backward algorithms + workspace(s)
-    const int kRequestedForwardAlgoCount = 6;
-    vector<cudnnConvolutionFwdAlgoPerf_t> fwd_perf;
-    fwd_perf.resize(kRequestedForwardAlgoCount);
-    int returnedAlgoCount;
-    CUDNN_CHECK(cudnnFindConvolutionForwardAlgorithm(handle_[0],
-                                                     bottom_descs_[i],
-                                                     filter_desc_,
-                                                     conv_descs_[i],
-                                                     top_descs_[i],
-                                                     kRequestedForwardAlgoCount,
-                                                     &returnedAlgoCount,
-                                                     &fwd_perf[0]));
-    layer_perf_.fwd_perf[i] =
-        vector<cudnnConvolutionFwdAlgoPerf_t>(fwd_perf.begin(), fwd_perf.begin() + returnedAlgoCount);
-
-
-    // choose backward algorithm for filter
-    const int kRequestedBackwardFilterAlgoCount = 4;
-    vector<cudnnConvolutionBwdFilterAlgoPerf_t> bwd_filter_perf;
-    bwd_filter_perf.resize(kRequestedBackwardFilterAlgoCount);
-    CUDNN_CHECK(cudnnFindConvolutionBackwardFilterAlgorithm(handle_[0],
-                                                            bottom_descs_[i],
+      // choose backward algorithm for filter
+      const int kRequestedBackwardFilterAlgoCount = 4;
+      vector<cudnnConvolutionBwdFilterAlgoPerf_t> bwd_filter_perf;
+      bwd_filter_perf.resize(kRequestedBackwardFilterAlgoCount);
+      CUDNN_CHECK(cudnnFindConvolutionBackwardFilterAlgorithm(handle_[0],
+                                                              bottom_descs_[i],
+                                                              top_descs_[i],
+                                                              conv_descs_[i],
+                                                              filter_desc_,
+                                                              kRequestedBackwardFilterAlgoCount,
+                                                              &returnedAlgoCount,
+                                                              &bwd_filter_perf[0]));
+
+      // choose the fastest within limit
+      // if all algorithms exceed memory limit, we will use the 0 algorithm with no workspace
+      for (int a = 0; a < kRequestedBackwardFilterAlgoCount; ++a){
+        if ((bwd_filter_perf[a].memory * this->group_ / 1024 / 1024) < Caffe::cudnn_mem_richness()
```

Comment:
Maybe better change to `bwd_filter_perf[a].memory * this->group_ <= Caffe::cudnn_mem_richness() * 1024 * 1024`?